### DISCUSSION TITLE
Getting a string value for Varnode which isConstant() == true (Java)

### DISCUSSION THREAD

**Discussion by Martyx00** [NationalSecurityAgency/ghidra#2650] [Q&A] [answered]
Hello,
I am struggling for couple hours on trying to get a String value (imagine for example format string such as `Hello %s!\n` for a function call to `printf`) after I get the `Varnode` (I get the `Varnode` by calling the `getInput(1)` on the `PcodeOP` of the `printf` function call) .  I am able to retrieve `Address` of that string in the binary with ease, however, I have not found a way how to read the String value of that constant itself.

Any help is greatly appreciated! Thanks!

**Comment by astrelsky** [comment, marked as answer]
Once you have the address of the string you may retrieve the `Data` using `getDataAt(Address)` or create it using `createAsciiString(Address)` if not already created. Once you have the `Data` the string value may be obtained via
`(String) data.getValue()`. The `Data` interface has a `hasStringValue` method if you wish to check before casting the value.

**Reply by Martyx00** [reply]
Thanks for a prompt reply!
I have tried to create the string with a call to `createAsciiString(Address)` from the `FlatProgramAPI` but I am getting a `ghidra.program.model.util.CodeUnitInsertionException: TerminatedCString may only be applied on initialized memory (const:00100864)` exception. Without this step, using a `getDataAt(Address)` either from the `FlatProgramAPI` or from `Listing` both returns `null`. It seems like I skipped some important step in the initialization of the `Program`. What would be a correct way to get initialized memory blocks?

**Reply by astrelsky** [reply]
How did you load the program?

**Reply by Martyx00** [reply]
I have created a new module in Eclipse and just used the `program` parameter that is passed to the pre-generated `added` function (Nothing else was done with it). Also, I am running the plugin after the basic analysis is completed.

**Reply by astrelsky** [reply]
> I have created a new module in Eclipse and just used the `program` parameter that is passed to the pre-generated `added` function (Nothing else was done with it). Also, I am running the plugin after the basic analysis is completed.

I mean when you added the file to the ghidra project, what loader was used?

There are a few situations where the address may be in dynamic memory. If this is the case then the string isn't a constant and you can just skip it.

**Reply by Martyx00** [reply]
It is a completely dummy ELF file I created to learn how to work with Ghidra API. Loading was done with standard ELF format file and `x86:LE:64:default:gcc` language. The basic analysis was left in the default state and the module was run afterwards as "One-shot". The string constant in question can be clearly seen in the `Listing` at the same address that I am trying to get it from (`0x00100864`) with `getDataAt`. It was also properly recognized as a `String` value by the decompiler. I have attached the file. 
[strcpy_test.zip](https://github.com/NationalSecurityAgency/ghidra/files/5834049/strcpy_test.zip)

**Reply by astrelsky** [reply]
It seems like there is some other issue occurring. May you post your code?

**Reply by dev747368** [reply]
I think the issue is that the address that you get from the varnode is in a different address space (const) and needs to be converted to the ram address space.  A recent commit (28e0480f9db47a41859abf5f20eb2dfc8dbeda4d) that added an analyzer that also pulls printf-style strings has an example of converting the varnode const address to ram addresses.

**Reply by Martyx00** [reply]
@dev747368 that works! Thanks!

**Reply by Heath123** [reply]
For anyone seeing this in the future here is the relevant code:
```java
private Address convertAddressToRamSpace(Address address) {

    String addressString = address.toString(false);
    return this.program.getAddressFactory().getAddress(addressString);
}
```
